sled-agent/src/rack_setup/service.rs

```diff
@@ -428,6 +440,85 @@ impl ServiceInner {
         Ok(())
     }
 
+    /// Requests that the specified sled configure storage as described
+    /// by `storage_config`.
+    ///
+    /// This function succeeds if either the configuration is supplied, or if
+    /// the configuration on the target sled is newer than what we're supplying.
+    // This function shares a lot of implementation details with
+    // [Self::initialize_zones_on_sled]. Although it has a different meaning,
+    // the usage (and expectations around generation numbers) are similar.
+    async fn initialize_datasets_on_sled(
+        &self,
+        sled_address: SocketAddrV6,
+        storage_config: DatasetsConfig,
+    ) -> Result<(), SetupServiceError> {
+        let dur = std::time::Duration::from_secs(60);
+        let client = reqwest::ClientBuilder::new()
+            .connect_timeout(dur)
+            .build()
+            .map_err(SetupServiceError::HttpClient)?;
+        let log = self.log.new(o!("sled_address" => sled_address.to_string()));
+        let client = SledAgentClient::new_with_client(
+            &format!("http://{}", sled_address),
+            client,
+            log.clone(),
+        );
+
+        let storage_put = || async {
+            info!(
+                log,
+                "attempting to set up sled's datasets: {:?}", storage_config,
+            );
+            let result = client.datasets_put(&storage_config.clone()).await;
+            let Err(error) = result else {
+                return Ok::<
+                    (),
+                    BackoffError<SledAgentError<SledAgentTypes::Error>>,
+                >(());
+            };
+
+            if let sled_agent_client::Error::ErrorResponse(response) = &error {
```

Comment:
I'm failing to see how this could happen during an initial rack setup. We won't handoff to nexus until after RSS, which we won't do until we finish calling this method. The only way this can fail in this scenario is if we call RSS again after the rack is already initialized. That is gated elsewhere though and we shouldn't even get here right?

Reply:
I think this may have been a result of copying "how we initialize zones" to "disks" and now to "datasets".

I've removed this in 6f6a87c2b29e823a6b831278591dd168fafecec1 - any conflict is now a hard error, for both disk and dataset initialization.